bindings/serialize.h

```diff
@@ -0,0 +1,246 @@
+class ProfileLocation : public ProtoField {
+ private:
+  uint64_t id;
+  uint64_t mappingID;
+  uint64_t address;
+  std::vector<Line> line;
+  bool isFolded;
+
+ public:
+  ProfileLocation(uint64_t id, uint64_t mappingID, uint64_t address,
+                  std::vector<Line> line, bool isFolded);
+
+  uint64_t getID();
+  uint64_t getMappingID();
+  uint64_t getAddress();
+  std::vector<Line> getLine();
+  bool getIsFolded();
+
+  virtual void encode(std::vector<char>* buffer) const;
+};
+
+// Corresponds to Sample defined in third_party/proto/profile.proto.
+class Sample : public ProtoField {
+ private:
+  std::vector<uint64_t> locationID;
+  std::vector<int64_t> value;
+  std::vector<Label> label;
+
+ public:
+  Sample(std::vector<uint64_t> locationID, std::vector<int64_t> value,
+         std::vector<Label> label);
+  std::vector<uint64_t> getLocationID();
+  std::vector<int64_t> getValue();
+  std::vector<Label> getLabel();
+  virtual void encode(std::vector<char>* buffer) const;
+};
+
+class Node;
+
+typedef std::tuple<int64_t, int64_t, int64_t, std::string> LocationKey;
+struct locationKeyHash : public std::unary_function<LocationKey, size_t> {
+  size_t operator()(const LocationKey& key) const {
+    result_type const h1(std::hash<int64_t>{}(std::get<0>(key)));
+    result_type const h2(std::hash<int64_t>{}(std::get<1>(key)));
+    result_type const h3(std::hash<int64_t>{}(std::get<2>(key)));
+    result_type const h4(std::hash<std::string>{}(std::get<3>(key)));
+    return h4 ^ (h3 << 1) ^ (h2 << 2) ^ (h1 << 4);
+  }
+};
+
+typedef std::tuple<int64_t, std::string> FunctionKey;
+struct functionKeyHash : public std::unary_function<FunctionKey, size_t> {
+  size_t operator()(const FunctionKey& key) const {
+    result_type const h1(std::hash<int64_t>{}(std::get<0>(key)));
+    result_type const h2(std::hash<std::string>{}(std::get<1>(key)));
+    return h2 ^ (h1 << 1);
+  }
+};
+
+// Corresponds to Profile defined in third_party/proto/profile.proto.
+class Profile : public ProtoField {
+ private:
+  std::vector<ValueType> sampleType;
+  std::vector<ProfileLocation> location;
+  std::vector<Sample> sample;
+  std::vector<Mapping> mapping;
+  std::vector<ProfileFunction> function;
+  std::vector<std::string> strings;
+  std::vector<int64_t> commentX;  // Indices into string table.
+  std::unordered_map<FunctionKey, int64_t, functionKeyHash> functionIDMap;
+  std::unordered_map<LocationKey, int64_t, locationKeyHash> locationIDMap;
+  std::unordered_map<std::string, int64_t> stringIDMap;
+  int64_t period;
+  int64_t timeNanos;
+  int64_t durationNanos;
+  int64_t defaultSampleTypeX;
+  ValueType periodType;
+  int64_t dropFramesX;  // Index into string table.
+  int64_t keepFramesX;  // Index into string table.
+
+ public:
+  Profile(std::string periodType, std::string periodUnit, int64_t period,
+          int64_t timeNanos = 0, int64_t durationNanos = 0,
+          std::string dropFrames = "", std::string keepFramesX = "");
```

Comment:
This is a large data structure. You should disallow copying and assignment to make sure there are no accidental copies that get generated. Likewise for other data-structures you define that have a non-trivial structure (e.g. have a list as a member).

Reply:
Done.